numpy/lib/_format_impl.py

```diff
@@ -684,13 +684,21 @@ def _read_array_header(fp, version, max_header_size=_MAX_HEADER_SIZE):
         raise ValueError(msg.format(keys))
 
     # Sanity-check the values.
-    if (not isinstance(d['shape'], tuple) or
-            not all(isinstance(x, int) for x in d['shape'])):
+    if not isinstance(d['shape'], tuple):
         msg = "shape is not valid: {!r}"
         raise ValueError(msg.format(d['shape']))
+
+    # Convert np.int64 values to Python int only after the check for tuple
+    d['shape'] = tuple(int(x) for x in d['shape'])
+
+    if not all(isinstance(x, int) for x in d['shape']):
+        msg = "shape contains non-integer values: {!r}"
+        raise ValueError(msg.format(d['shape']))
+
```

Comment:
Why change this here.  What is stored must be integers already, so this is fine as is.

(Also the isinstance check now can never fail!)